packages/contracts-bedrock/src/Safe/LivenessGuard.sol

```diff
@@ -0,0 +1,141 @@
+        address[] memory ownersAfter = SAFE.getOwners();
+
+        // Iterate over the current owners, and remove one at a time from the ownersBefore set.
+        uint256 ownersAfterLength = ownersAfter.length;
+        for (uint256 i = 0; i < ownersAfterLength; i++) {
+            // If the value was present, remove() returns true.
+            address ownerAfter = ownersAfter[i];
+            if (ownersBefore.remove(ownerAfter) == false) {
+                // This address was not already an owner, add it to the lastLive mapping
+                lastLive[ownerAfter] = block.timestamp;
+            }
+        }
+        // Now iterate over the remaining ownersBefore entries. Any remaining addresses are no longer an owner, so we
+        // delete them from the lastLive mapping.
+        for (uint256 j = 0; j < ownersBefore.length(); j++) {
+            address ownerBefore = ownersBefore.at(j);
+            delete lastLive[ownerBefore];
+        }
+    }
+
+    /// @notice Records the most recent time which any owner has signed a transaction.
+    /// @dev This method is called by the Safe contract, it is critical that it does not revert, otherwise
+    ///      the Safe contract will be unable to execute transactions.
+    function checkTransaction(
+        address to,
+        uint256 value,
+        bytes memory data,
+        Enum.Operation operation,
+        uint256 safeTxGas,
+        uint256 baseGas,
+        uint256 gasPrice,
+        address gasToken,
+        address payable refundReceiver,
+        bytes memory signatures,
+        address
+    )
+        external
+    {
+        require(msg.sender == address(SAFE), "LivenessGuard: only Safe can call this function");
+
+        // Cache the set of owners prior to execution.
+        // This will be used in the checkAfterExecution method.
+        address[] memory owners = SAFE.getOwners();
+        for (uint256 i = 0; i < owners.length; i++) {
+            ownersBefore.add(owners[i]);
+        }
+
+        // This call will reenter to the Safe which is calling it. This is OK because it is only reading the
+        // nonce, and using the getTransactionHash() method.
+        bytes32 txHash = Safe(payable(msg.sender)).getTransactionHash({
+            to: to,
+            value: value,
+            data: data,
+            operation: operation,
+            safeTxGas: safeTxGas,
+            baseGas: baseGas,
+            gasPrice: gasPrice,
+            gasToken: gasToken,
+            refundReceiver: refundReceiver,
+            _nonce: Safe(payable(msg.sender)).nonce() - 1
+        });
+
+        uint256 threshold = SAFE.getThreshold();
+        address[] memory signers =
+            SafeSigners.getNSigners({ dataHash: txHash, signatures: signatures, requiredSignatures: threshold });
+
+        for (uint256 i = 0; i < signers.length; i++) {
+            lastLive[signers[i]] = block.timestamp;
+        }
+        emit SignersRecorded(txHash, signers);
+    }
+
```

Comment:
Also we use "owners" again here and in the natspec above, but we should probably be consistent throughout
```suggestion
        require(safe.isOwner(msg.sender), "LivenessGuard: only Safe owners may demonstrate liveness");
```

Reply:
This is the one place where I think `signers` is a better term, because these values are being returned by `SafeSigners.getNSigners()`, and `getNOwners` wouldn't be an appropriate name for that method IMO.

Is that OK if kept isolated to this location?

Reply:
Oops that suggestion got messed up, was trying to fix the typo in `demontstrate` there

> This is the one place where I think `signers` is a better term, because these values are being returned by `SafeSigners.getNSigners()`, and `getNOwners` wouldn't be an appropriate name for that method IMO.
> 
> Is that OK if kept isolated to this location?

Yep that seems ok to me